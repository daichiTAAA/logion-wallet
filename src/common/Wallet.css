.Wallet .asset-name-cell {
    padding-top: 3px;
}

.Wallet .asset-name-cell .name {
    margin-left: 10px;
}

.Wallet .not-available {
    display: flex;
    align-items: center;
    justify-content: center;
    color: #e6007a;
    border: 2px solid #e6007a;
    border-radius: 6px;
    padding: 9px 20px;
}

.Wallet .gauge-title {
    display: flex;
    justify-content: center;
    margin-bottom: 40px;
    align-items: center;
}

.Wallet .gauge-title img {
    margin-right: 10px;
}

