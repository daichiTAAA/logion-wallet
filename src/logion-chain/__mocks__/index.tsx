import { Extension } from '../Keys';

const LogionChainContextProvider = (props: any) => null;

let context = {
    apiState: 'CONNECT_INIT',
    injectedAccounts: []
};

export function setContextMock(value: any) {
    context = value;
}

export function useLogionChain() {
    return context;
}

let extensionAvailable = false;

export function setExtensionAvailable(value: boolean) {
    extensionAvailable = value;
}

export function isExtensionAvailable() {
    return extensionAvailable;
}

let recommendedExtensionValue: Extension | null = null;

export function setRecommendedExtension(value: Extension | null) {
    recommendedExtensionValue = value;
}

export function recommendedExtension(): Extension | null {
    return recommendedExtensionValue;
}

export {
    LogionChainContextProvider,
};
