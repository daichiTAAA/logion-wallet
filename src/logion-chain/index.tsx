// Imports
import { KeyringPair as PolkadotKeyringPair } from '@polkadot/keyring/types';
import {
    Block as PolkadotBlock,
    Hash as PolkadotHash,
    Header as PolkadotHeader,
    SignedBlock as PolkadotSignedBlock,
    EventRecord as PolkadotEventRecord,
    Extrinsic as PolkadotExtrinsic,
} from '@polkadot/types/interfaces';
import keyring from '@polkadot/ui-keyring';
import { InjectedAccountWithMeta as PolkadotInjectedAccountWithMeta } from '@polkadot/extension-inject/types';
import { ISubmittableResult as PolkadotISubmittableResult } from '@polkadot/types/types';

import {
    LogionChainContextProvider,
    useLogionChain,

    ExtrinsicFetchSpecification as ContextExtrinsicFetchSpecification,
    ApiState as ContextApiState,
    NodeMetadata as ContextNodeMetadata,
} from './LogionChainContext';

import {
    signAndSend,
    replaceUnsubscriber,
    sign,

    ExtrinsicSignatureParameters as SignatureExtrinsicSignatureParameters,
    Unsubscriber as SignatureUnsubscriber,
    SignAndSendCallback as SignatureSignAndSendCallback,
    AttributesSignatureParameters as SignatureStringSignatureParameters,
} from './Signature';

import {
    getTimestamp,
    fetchExtrinsics,
    hashEquals,

    ExtrinsicsAndHead as BlocksExtrinsicsAndHead,
} from './Blocks';

import {
    isExtensionAvailable,
    recommendedExtension
} from './Keys';

// Re-exports
export {
    useLogionChain,
    LogionChainContextProvider,
    getTimestamp,
    fetchExtrinsics,
    hashEquals,
    signAndSend,
    replaceUnsubscriber,
    isExtensionAvailable,
    recommendedExtension,
    keyring,
    sign,
};

export type Unsubscriber = SignatureUnsubscriber;
export type ExtrinsicSignatureParameters = SignatureExtrinsicSignatureParameters;
export type ExtrinsicFetchSpecification = ContextExtrinsicFetchSpecification;

export type ExtrinsicsAndHead = BlocksExtrinsicsAndHead;

export type KeyringPair = PolkadotKeyringPair;
export type Hash = PolkadotHash;
export type SignedBlock = PolkadotSignedBlock;
export type Header = PolkadotHeader;
export type Block = PolkadotBlock;
export type EventRecord = PolkadotEventRecord;
export type Extrinsic = PolkadotExtrinsic;
export type SignAndSendCallback = SignatureSignAndSendCallback;
export type ApiState = ContextApiState;
export type NodeMetadata = ContextNodeMetadata;
export type InjectedAccountWithMeta = PolkadotInjectedAccountWithMeta;
export type StringSignatureParameters = SignatureStringSignatureParameters;
export type ISubmittableResult = PolkadotISubmittableResult;
