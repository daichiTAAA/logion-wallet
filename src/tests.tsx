import { ReactElement } from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import reactRenderer, { ReactTestRenderer, ReactTestRendererJSON, act as reactAct } from 'react-test-renderer';

export function render(element: ReactElement): null | ReactTestRendererJSON | ReactTestRendererJSON[] {
    const component = reactRenderer.create(element);
    return component.toJSON();
}

export function renderer(element: ReactElement): ReactTestRenderer {
    return reactRenderer.create(element);
}

export function shallowRender<E extends ReactElement>(element: ReactElement): E {
    const renderer = createRenderer();
    renderer.render(element);
    return renderer.getRenderOutput();
}

export function mockAccount(address: string, name: string) {
    return {
        address,
        meta: {
            name
        }
    };
}

export { reactAct as act };
