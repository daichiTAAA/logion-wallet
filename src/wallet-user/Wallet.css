.Wallet .actions {
    display: flex;
    justify-content: space-between;
    padding-left: 130px;
    padding-right: 130px;
    margin-top: 60px;
    margin-bottom: 30px;
}

.Wallet .assets {
    margin-top: 40px;
}

.Wallet .asset-name-cell {
    padding-top: 3px;
}

.Wallet .asset-name-cell .name {
    margin-left: 10px;
}
