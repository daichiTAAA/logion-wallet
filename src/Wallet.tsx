import React from 'react';
import Button from 'react-bootstrap/Button';
import Jumbotron from 'react-bootstrap/Jumbotron';
import { useLogionChain, ApiState, NodeMetadata } from './logion-chain';

import Shell from './Shell';

function status(apiState: ApiState, metadata: NodeMetadata | null): string {
    if(apiState === 'READY' && metadata === null) {
        return "connected";
    } else if(apiState === 'READY' && metadata !== null) {
        return `connected to node "${metadata.name}" (Peer ID: ${metadata.peerId})`;
    } else if(apiState === 'DISCONNECTED') {
        return "disconnected";
    } else if(apiState === 'ERROR') {
        return "disconnected (error)";
    } else {
        return "connecting";
    }
}

export default function Wallet() {
    const { injectedAccounts, apiState, connect, connectedNodeMetadata } = useLogionChain();
    if(injectedAccounts === null) {
        return null;
    }

    const connectButton = apiState === 'DISCONNECTED' ? <Button onClick={connect}>Connect</Button> : null;

    return (
        <Shell>
            <Jumbotron>
                <h1>You are ready to use the Logion wallet, congratulations!</h1>
            </Jumbotron>

            <p>The following accounts were detected:</p>
            <ul>
                {injectedAccounts.map(injectedAccount => <li key={injectedAccount.address}>{injectedAccount.address} ({injectedAccount.meta.name || ""})</li>)}
            </ul>
            <p>You are currently {status(apiState, connectedNodeMetadata)}</p>
            {connectButton}
        </Shell>
    );
}
