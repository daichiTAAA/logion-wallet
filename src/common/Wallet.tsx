import React from 'react';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import { useHistory } from 'react-router-dom';

import { CoinBalance, prefixedLogBalance } from '../logion-chain/Balances';

import { useCommonContext } from './CommonContext';

import { FullWidthPane } from './Dashboard';
import Frame from './Frame';
import Table, { Cell, DateCell, EmptyTableMessage, ActionCell } from './Table';
import Icon from './Icon';
import Button from './Button';
import WalletGauge from './WalletGauge';
import Loader from './Loader';
import { Transaction } from './types/ModelTypes';

import './Wallet.css';

export interface Props {
    transactionsPath: (coinId: string) => string,
}

export default function Wallet(props: Props) {
    const { colorTheme } = useCommonContext();

    return (
        <FullWidthPane
            className="Wallet"
            mainTitle="Wallet"
            titleIcon={{
                icon: {
                    id: 'wallet'
                },
                background: colorTheme.topMenuItems.iconGradient,
            }}
        >
            <Content { ...props } />
        </FullWidthPane>
    );
}

export function Content(props: Props) {
    const { balances, transactions, colorTheme } = useCommonContext();
    const history = useHistory();

    if(balances === null || transactions === null) {
        return <Loader />;
    }

    const latestTransaction = transactions[0];

    return (
        <>
            <Row>
                <Col md={8}>
                    <Frame
                        title="Balance per month"
                    >
                        <img
                            className="fake-graph"
                            src={ `${process.env.PUBLIC_URL}/assets/themes/${colorTheme.type}/fake_balance_history.png` }
                            alt="Fake balance graph"
                        />
                    </Frame>
                </Col>
                <Col md={4}>
                    <Frame
                        title="Current LOG balance"
                        fillHeight
                    >
                        <WalletGauge
                            coin={ balances[0].coin }
                            balance={ balances[0].balance }
                            level={ balances[0].level }
                            type='arc'
                        />
                    </Frame>
                </Col>
            </Row>
            <Frame
                className="assets"
            >
                <h2>Asset balances</h2>

                <Table
                    columns={[
                        {
                            header: "Asset name",
                            render: balance => <AssetNameCell balance={ balance } />,
                            align: 'left',
                        },
                        {
                            header: "Balance",
                            render: balance => <Cell content={ balance.balance.coefficient.toFixedPrecision(2) } />,
                            width: "150px",
                            align: 'right',
                        },
                        {
                            header: "Last transaction date",
                            render: balance => <DateCell dateTime={ balance.coin.id !== 'dot' && latestTransaction !== undefined ? latestTransaction.createdOn : null } />
                        },
                        {
                            header: "Last transaction type",
                            render: balance => <Cell content={ balance.coin.id !== 'dot' && latestTransaction !== undefined ? latestTransaction.type : "-" } />
                        },
                        {
                            header: "Last transaction amount",
                            render: balance => <Cell content={ balance.coin.id !== 'dot' && latestTransaction !== undefined ? prefixedLogBalance(transactionAmount(latestTransaction)).convertTo(balance.balance.prefix).coefficient.toFixedPrecision(2) : '-' } />,
                            align: 'right',
                        },
                        {
                            header: "",
                            render: balance => balance.coin.id !== 'dot' ? <ActionCell><Button onClick={() => history.push(props.transactionsPath(balance.coin.id))}>More</Button></ActionCell> : <NotAvailable/>,
                            width: "200px",
                        }
                    ]}
                    data={ balances }
                    renderEmpty={ () => <EmptyTableMessage>You have no asset yet</EmptyTableMessage> }
                />
            </Frame>
        </>
    );
}

interface AssetNameCellProps {
    balance: CoinBalance,
}

function AssetNameCell(props: AssetNameCellProps) {

    return (
        <div className="asset-name-cell">
            <Icon icon={{id: props.balance.coin.iconId}} type={ props.balance.coin.iconType } height="36px" width="auto" />
            <span className="name">{ props.balance.coin.name } ({ props.balance.balance.prefix.symbol }{ props.balance.coin.symbol })</span>
        </div>
    );
}

function NotAvailable() {

    return (
        <div className="not-available">
            <span>Not available (yet)</span>
        </div>
    );
}

function transactionAmount(transaction: Transaction): string {
    if(transaction.type === 'Received') {
        return transaction.transferValue;
    } else if(transaction.type === 'Sent') {
        return "-" + transaction.transferValue;
    } else {
        return transaction.total;
    }
}
