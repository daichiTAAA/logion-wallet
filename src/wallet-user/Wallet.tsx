import React, { useEffect, useState } from 'react';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';

import { useLogionChain } from '../logion-chain';
import { AccountData, getAccountData, LOG_DECIMALS } from '../logion-chain/Balances';
import { ScientificNumber, PrefixedNumber, convertToPrefixed } from '../logion-chain/numbers';

import { FullWidthPane } from '../common/Dashboard';
import Frame from '../common/Frame';
import Gauge from '../common/Gauge';
import MenuIcon from '../common/MenuIcon';
import Table, { Cell, DateCell, EmptyTableMessage } from '../common/Table';
import Button from '../common/Button';
import Clickable from '../common/Clickable';
import Icon from '../common/Icon';
import { useRootContext } from '../RootContext';

import { useUserContext } from "./UserContext";

import './Wallet.css';

const ARTIFICIAL_MAX_BALANCE = new ScientificNumber("100", -LOG_DECIMALS);

export default function Account() {
    const { selectAddress, addresses } = useRootContext();
    const { colorTheme } = useUserContext();
    const { api } = useLogionChain();
    const [ dataAddress, setDataAddress ] = useState<string | null>(null);
    const [ accountData, setAccountData ] = useState<AccountData | null>(null);
    const [ balance, setBalance ] = useState<PrefixedNumber>(PrefixedNumber.ZERO);
    const [ level, setLevel ] = useState<number>(0);

    useEffect(() => {
        if(addresses !== null
            && (accountData === null || addresses.currentAddress.address !== dataAddress)) {
            setDataAddress(addresses.currentAddress.address);
            setBalance(PrefixedNumber.ZERO);
            setLevel(0);
            (async function() {
                const data = await getAccountData({
                    api: api!,
                    accountId: addresses.currentAddress.address,
                });
                setAccountData(data);
                const available = new ScientificNumber(data.available, -LOG_DECIMALS).optimizeScale(3);
                setBalance(convertToPrefixed(available));
                setLevel(available.divideBy(ARTIFICIAL_MAX_BALANCE).toNumber());
            })();
        }
    }, [ accountData, addresses, api, dataAddress ]);

    if(addresses === null || selectAddress === null) {
        return null;
    }

    let allAssets = [
        {
            name: "LOG",
            iconId: 'log',
            balance: `${balance.coefficient.toFixedPrecision(2)} ${balance.prefix.symbol}LOG`,
            lastTransactionDate: null,
            lastTransactionType: null,
            lastTransactionAmount: null,
        },
        {
            name: "DOT",
            iconId: 'dot',
            balance: `0.00 DOT`,
            lastTransactionDate: null,
            lastTransactionType: null,
            lastTransactionAmount: null,
        }
    ];

    return (
        <FullWidthPane
            className="Wallet"
            mainTitle="Wallet"
            titleIcon={{
                icon: {
                    id: 'wallet'
                },
                background: colorTheme.topMenuItems.iconGradient,
            }}
            colors={ colorTheme }
            addresses={ addresses }
            selectAddress={ selectAddress }
        >
            <Row>
                <Col md={4}>
                    <Frame
                        colors={ colorTheme }
                    >
                        <Gauge
                            title="Current LOG balance"
                            readingIntegerPart={ balance.coefficient.toInteger() }
                            readingDecimalPart={ balance.coefficient.toFixedPrecisionDecimals(2) }
                            unit={ balance.prefix.symbol + "LOG" }
                            level={ level }
                        />
                        <div className="actions">
                            <Clickable>
                                <MenuIcon
                                    icon={{
                                        id: 'send'
                                    }}
                                    background={ colorTheme.topMenuItems.iconGradient }
                                    colorThemeType={ colorTheme.type }
                                />
                            </Clickable>
                            <Clickable>
                                <MenuIcon
                                    icon={{
                                        id: 'buy'
                                    }}
                                    background={ colorTheme.topMenuItems.iconGradient }
                                    colorThemeType={ colorTheme.type }
                                />
                            </Clickable>
                        </div>
                    </Frame>
                </Col>
            </Row>
            <Frame
                className="assets"
                colors={ colorTheme }
            >
                <h2>Asset balances</h2>

                <Table
                    colorTheme={ colorTheme }
                    columns={[
                        {
                            header: "Asset name",
                            render: asset => <AssetNameCell asset={ asset } />
                        },
                        {
                            header: "Balance",
                            render: asset => <Cell content={ asset.balance } align="right"/>,
                            width: "150px",
                        },
                        {
                            header: "Last transaction date",
                            render: asset => <DateCell dateTime={ asset.lastTransactionDate } />
                        },
                        {
                            header: "Last transaction type",
                            render: asset => <Cell content={ asset.lastTransactionType } align="center"/>
                        },
                        {
                            header: "Last transaction amount",
                            render: asset => <Cell content={ asset.lastTransactionAmount } align="right"/>
                        },
                        {
                            header: "",
                            render: asset => <Button colors={ colorTheme.buttons }>More</Button>
                        }
                    ]}
                    data={ allAssets }
                    renderEmpty={ () => <EmptyTableMessage>You have no asset yet</EmptyTableMessage> }
                />
            </Frame>
        </FullWidthPane>
    );
}

interface Asset {
    name: string,
    iconId: string,
    balance: string,
    lastTransactionDate: string | null,
    lastTransactionType: string | null,
    lastTransactionAmount: string | null,
}

interface AssetNameCellProps {
    asset: Asset,
}

function AssetNameCell(props: AssetNameCellProps) {

    return (
        <div className="asset-name-cell">
            <Icon icon={{id: props.asset.iconId}} type="png" />
            <span className="name">{ props.asset.name }</span>
        </div>
    );
}
