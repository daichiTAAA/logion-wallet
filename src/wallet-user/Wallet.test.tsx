jest.mock('../logion-chain');
jest.mock('./UserContext');

import React from 'react';
import Wallet from './Wallet';
import { setContextMock } from '../logion-chain';
import { shallowRender, mockAccount } from '../tests';

test('Given disconnected, when rendering, then show disconnected and connect button', () => {
    setContextMock({
        apiState: 'DISCONNECTED',
        injectedAccounts: [
            mockAccount("address", "name")
        ]
    });

    const tree = shallowRender(<Wallet />);

    expect(tree).toMatchSnapshot();
});

test('Given connected, when rendering, then show account, no connect button and tokenization', () => {
    setContextMock({
        apiState: 'READY',
        injectedAccounts: [
            mockAccount("address", "Account name")
        ],
        connectedNodeMetadata: {
            name: "Node name",
            peerId: "Node peer ID"
        }
    });

    const tree = shallowRender(<Wallet />);

    expect(tree).toMatchSnapshot();
});

test('Given connected and no metadata, when rendering, then show account, no connect button, no node info, tokenization', () => {
    setContextMock({
        apiState: 'READY',
        injectedAccounts: [
            mockAccount("address", "Account name")
        ],
        connectedNodeMetadata: null
    });

    const tree = shallowRender(<Wallet />);

    expect(tree).toMatchSnapshot();
});

test('Given connection error, when rendering, then show account, no button, no node info', () => {
    setContextMock({
        apiState: 'ERROR',
        injectedAccounts: [
            mockAccount("address", "Account name")
        ],
        connectedNodeMetadata: null
    });

    const tree = shallowRender(<Wallet />);

    expect(tree).toMatchSnapshot();
});

test('Given connecting, when rendering, then show account, no button, no node info', () => {
    setContextMock({
        apiState: 'CONNECTING',
        injectedAccounts: [
            mockAccount("address", "Account name")
        ],
        connectedNodeMetadata: null
    });

    const tree = shallowRender(<Wallet />);

    expect(tree).toMatchSnapshot();
});

test('Given no injected account, when rendering, then nothing displayed', () => {
    setContextMock({
        apiState: 'READY',
        injectedAccounts: null,
        connectedNodeMetadata: null
    });

    const tree = shallowRender(<Wallet />);

    expect(tree).toMatchSnapshot();
});
