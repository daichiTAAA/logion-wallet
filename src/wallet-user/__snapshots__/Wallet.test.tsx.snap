// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given connected and no metadata, when rendering, then show account, no connect button, no node info, tokenization 1`] = `
<Shell>
  <h1>
    You are ready to use the Logion wallet, congratulations!
  </h1>
  <p>
    The following account(s) were detected:
  </p>
  <ul>
    <li>
      <Link
        to="/user/tokens/address"
      >
        address
         (
        Account name
        )
      </Link>
    </li>
  </ul>
  <p>
    You are currently 
    connected
  </p>
  <UserContextProvider
    legalOfficerAddress="5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"
    userAddress="address"
  >
    <UserRouter />
  </UserContextProvider>
</Shell>
`;

exports[`Given connected, when rendering, then show account, no connect button and tokenization 1`] = `
<Shell>
  <h1>
    You are ready to use the Logion wallet, congratulations!
  </h1>
  <p>
    The following account(s) were detected:
  </p>
  <ul>
    <li>
      <Link
        to="/user/tokens/address"
      >
        address
         (
        Account name
        )
      </Link>
    </li>
  </ul>
  <p>
    You are currently 
    connected to node "Node name" (Peer ID: Node peer ID)
  </p>
  <UserContextProvider
    legalOfficerAddress="5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"
    userAddress="address"
  >
    <UserRouter />
  </UserContextProvider>
</Shell>
`;

exports[`Given connecting, when rendering, then show account, no button, no node info 1`] = `
<Shell>
  <h1>
    You are ready to use the Logion wallet, congratulations!
  </h1>
  <p>
    The following account(s) were detected:
  </p>
  <ul>
    <li>
      <Link
        to="/user/tokens/address"
      >
        address
         (
        Account name
        )
      </Link>
    </li>
  </ul>
  <p>
    You are currently 
    connecting
  </p>
</Shell>
`;

exports[`Given connection error, when rendering, then show account, no button, no node info 1`] = `
<Shell>
  <h1>
    You are ready to use the Logion wallet, congratulations!
  </h1>
  <p>
    The following account(s) were detected:
  </p>
  <ul>
    <li>
      <Link
        to="/user/tokens/address"
      >
        address
         (
        Account name
        )
      </Link>
    </li>
  </ul>
  <p>
    You are currently 
    disconnected (error)
  </p>
</Shell>
`;

exports[`Given disconnected, when rendering, then show disconnected and connect button 1`] = `
<Shell>
  <h1>
    You are ready to use the Logion wallet, congratulations!
  </h1>
  <p>
    The following account(s) were detected:
  </p>
  <ul>
    <li>
      <Link
        to="/user/tokens/address"
      >
        address
         (
        name
        )
      </Link>
    </li>
  </ul>
  <p>
    You are currently 
    disconnected
  </p>
  <Button
    active={false}
    disabled={false}
    variant="primary"
  >
    Connect
  </Button>
</Shell>
`;

exports[`Given no injected account, when rendering, then nothing displayed 1`] = `null`;
